NXP S32 introduce support SENT

congnguyenhuu opened the pull request:
Introduce support Single Edge Nibble Transmission (SENT) driver using peripheral SENT receiver on S32Z. This driver allows to communication (read data) with SENT device.
The test result: [SENT-RECEIVER-TESTING-RESULT.pptx](https://github.com/user-attachments/files/17946312/SENT-RECEIVER-TESTING-RESULT.pptx)

RFC: https://github.com/zephyrproject-rtos/zephyr/issues/83983

congnguyenhuu committed:
drivers: introduce Single Edge Nibble Transmission (SENT) driver

This driver allows to communication (receive data) with SENT device

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
boards: s32z270: enable support SENT

enable support SENT

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
tests/samples: drivers: create test, sample for SENT driver

Create test, sample for SENT driver

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

manuargue commented:
@congnguyenhuu I'd suggest you to open an RFC (request for comments) and link it to this pr in order to involve community to review the proposal and reach consensus: https://docs.zephyrproject.org/latest/contribute/proposals_and_rfcs.html

JarmouniA reviewed (commented):
I see the same concerns raised in the comments  https://github.com/zephyrproject-rtos/zephyr/pull/79824#discussion_r1916764401 and https://github.com/zephyrproject-rtos/zephyr/pull/79824#pullrequestreview-2552800277 in this PR.

congnguyenhuu commented:
> I see the same concerns raised in the comments [#79824 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/79824#discussion_r1916764401) and [#79824 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/79824#pullrequestreview-2552800277) in this PR.

I addressed

rruuaanng commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,183 @@
+/**
+ * @brief SENT frame structure
+ */
+struct sent_frame {
```
Forgive me for not being very familiar with SENT, but it doesn't look very nice, is this the final solution?
(It's nested too much.)

congnguyenhuu replied:
yes, it's not pretty, not final. I will improve

congnguyenhuu replied:
I improved the frame layout, merged the common fields

manuargue replied:
`frame.data` vs `frame.slow_frame.data` doesn't read nice, the first could also be understood as slow data. Right now to interpret the frame, it's necessary the information obtained from the callback via `enum sent_status`.

With the idea of keeping a single frame model for fast and slow frames, consider adding a frame type indicator and making fast and slow data self contained. This also can give space to add other frame types in the future. Example:
```c
enum sent_frame_type {
    SENT_FAST_FRAME,
    SENT_SLOW_FRAME
};

struct sent_frame {
    enum sent_frame_type type;

    union {
        struct {
            uint16_t id;      
            uint16_t data;  
        } slow;

        struct {
            uint32_t data;  
        } fast;
    };

    uint32_t timestamp;
    uint8_t crc;
};
```

congnguyenhuu replied:
Thanks for suggestion, I updated

decsny commented:
This will need to go to architecture review when ready

congnguyenhuu commented:
Hello @henrikbrixandersen, I saw that my RFC is still listed under "Todo Architecture Review" status, no actions. Could you please help me expedite the review process? Thanks.

manuargue commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,179 @@
+enum sent_status {
+	SENT_RX_SLOW_FRAME,
+	SENT_RX_FAST_FRAME,
+	SENT_RX_ERR,
```
With this design there's no way to distinguish between errors in slow channels and fast channels. I'd imagine it would require different level of actions whether the error occurs in one or the other channel, e.g. fast channel might be a critical impact whereas slow channel might affect only diagnostics.

congnguyenhuu replied:
I separate 2 specific error statuses for serial and fast frame

congnguyenhuu replied:
Due to there are 2 specific callback for serial and fast frame so we can distinguish between errors

manuargue commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,179 @@
+				      sent_rx_callback_t callback, void *user_data);
+
+__subsystem struct sent_driver_api {
+	sent_start_rx_t start_rx;
+	sent_stop_rx_t stop_rx;
+	sent_add_rx_callback_t add_rx_callback;
```
Having "sent" and "rx" together in the function names can be very confusing and might be misinterpreted. Possible name choices:

- sent_listen() / sent_start_listening() / sent_start_capture()
- sent_stop() / sent_stop_listening() / sent_stop_capture()
- sent_register_callback()

I'm assuming you decided to use `_rx_` because SENT is unidirectional. However, when (and if) SENT-B would be available, this API will probably might need to be extended to support bi-directional communication so the "rx" may not be relevant then.

congnguyenhuu replied:
I replaced all names of the `sent_ ` prefix in the variations with `saej2716_`
with APIs: `saej2716_start_rx()`, `saej2716_stop_rx() `and `saej2716_add_rx_callback()`
I will consider more your suggestion

manuargue commented on drivers/sent/Kconfig:
```diff
@@ -0,0 +1,23 @@
+	help
+	  Enable SENT Driver Configuration
+
+if SENT
```
How about `CONFIG_SAEJ2716`? Similarly to `IEEE802154`, `IEEE80211*`, etc.

congnguyenhuu replied:
I replaced `SENT` to `SAEJ2716`

congnguyenhuu commented on include/zephyr/drivers/sent/sent.h:
thanks for your suggestion, I separated 2 specific callbacks for fast frame and serial frame

congnguyenhuu commented on samples/drivers/sent/src/main.c:
```diff
@@ -0,0 +1,36 @@
+void rx_cb(const struct device *dev, uint8_t channel_id, struct sent_frame *frame,
+	   enum sent_status status, void *user_data)
+{
+	LOG_INF("Rx channel %d completed\n", channel_id);
```
I updated to add more information

congnguyenhuu commented on samples/drivers/sent/src/main.c:
```diff
@@ -0,0 +1,36 @@
+
+	sent_start_rx(dev, SENT_CHANNEL);
+
+	k_sleep(K_MSEC(100));
```
I added an infinite loop for waiting to receive

congnguyenhuu commented:
Based on the suggestions from the Architecture WG meeting on March 4th, I have updated to the v1 changes. These updates include:
1. Changed `add_rx_callback(dev, channel, serial_cb, fast_cb, user_data)` to
`register_callback(dev, channel, struct saej2716_rx_callback_configs callback_configs)`
 
With the newly defined structure for RX callback configurations in `saej2716.h`
```
struct saej2716_rx_callback_config {
	saej2716_rx_frame_callback_t callback;
	struct saej2716_frame *frame;
	uint32_t max_num_frame;
	void *user_data;
};
 
struct saej2716_rx_callback_configs {
	struct saej2716_rx_callback_config *serial;
	struct saej2716_rx_callback_config *fast;
};
```
 
2. RX buffers are allocated by the user to improve throughput:
- Use the API `register_callback` to register the RX callback and allocate a user-defined RX buffer for each callback.
- Count the number of RX buffer frames and trigger the RX callback when the count reaches the maximum buffer frame limit.
- Reset the count when a RX error occurs or when the count reaches the maximum buffer frame limit.
 
3. Tests and samples have been updated to demonstrate usage with the new changes

Dat-NguyenDuy commented on dts/bindings/saej2716/saej2716-controller.yaml:
```diff
@@ -0,0 +1,80 @@
+      description: |
+        Indicates whether the channel is enabled to detect a pause pulse.
+
+    crc-check-disable:
```
For CRC, as far as i understand:
- Fast Message: CRC is optional (can be enabled/disabled). If CRC is enabled, two main options:
   + Legacy 4-bit CRC
   + Recommended 4-bit CRC
- Short Serial Message: CRC seems always enabled, two options:
   + Legacy 4-bit CRC
   + Recommended 4-bit CRC
- Enhanced Serial Message: 6-bit CRC

Should we do something like:
- In dts bindings, create two properties for fast-crc and short-serial-crc
- Define macros for these options in a header file, which can be combined to configure the CRC

congnguyenhuu replied:
I updated as your suggestion. added 2 properties `fast-crc` and `short-serial-crc`

Dat-NguyenDuy reviewed (commented):
Some comments regarding the DTS bindings and the API to be taken into account

congnguyenhuu commented on dts/bindings/saej2716/saej2716-controller.yaml:
```diff
@@ -0,0 +1,80 @@
+
+  properties:
+    reg:
+      type: array
```
I updated

congnguyenhuu commented on dts/bindings/saej2716/nxp,s32-saej2716.yaml:
```diff
@@ -0,0 +1,40 @@
+
+child-binding:
+
+  compatible: "nxp,s32-saej2716-channel"
```
yes, it is not needed. I removed

congnguyenhuu commented on dts/bindings/saej2716/nxp,s32-saej2716.yaml:
```diff
@@ -0,0 +1,40 @@
+        - 1024
+        - 2048
+      description: |
+        Specifies the number of bus timeout cycles. This value determines the maximum number
```
I added more information in description

congnguyenhuu commented on include/zephyr/drivers/saej2716/saej2716.h:
```diff
@@ -0,0 +1,206 @@
+/**
+ * @brief SAEJ2716 status
+ */
+enum saej2716_status {
```
yes, I agree. I updated the 
```
enum saej2716_status {
	SAEJ2716_RX_FRAME,
	SAEJ2716_RX_ERR,
};
```

Dat-NguyenDuy replied:
Looking into current implementation, the callback will be invoked when either:
- The number of (non-error) received frames is equal to the expected amount
- an error frame is received (number received frames != expected amount), only report non-error frame

This seems redundant to me, as the application can rely on the `num_frame` from the callback function to determine if there is an error.

Alternative solution to be consider is that, the callback will only be invoked when the number of received frames (including error frame) is equal to the expected amount, you will need to include status for each frame. The status could be: success, crc_error, etc

congnguyenhuu replied:
yes, I updated to remove `enum saej2716_status`. just report the `num_frame`.
An error frame is received (number received frames != expected amount), only report non-error frame

congnguyenhuu commented on include/zephyr/drivers/saej2716/saej2716.h:
```diff
@@ -0,0 +1,206 @@
+ * @param channel    The hardware channel of the driver instance.
+ * @param status     SAEJ2716 frame status.
+ * @param num_frame  Number of received frame.
+ * @param user_data  User data provided when the filter was added.
```
I updated

congnguyenhuu commented on dts/bindings/saej2716/saej2716-controller.yaml:
```diff
@@ -0,0 +1,80 @@
+        Specifies the number of data nibbles. A nibble is a four-bit aggregation.
+        This property defines how many such nibbles are used in the data transmission.
+
+    tick-time-prescaler-us:
```
I updated

congnguyenhuu commented on include/zephyr/drivers/saej2716/saej2716.h:
```diff
@@ -0,0 +1,206 @@
+ * @brief SAEJ2716 frame type
+ */
+enum saej2716_frame_type {
+	SAEJ2716_SERIAL_FRAME,
```
yes, I'm using a common type for both. serial frame include formats:
- short frame,
- enhanced frame 4 bit ID
- enhanced frame 8 bit ID
with new frame type

```
enum saej2716_frame_type {
	SAEJ2716_SHORT_SERIAL_FRAME,
	SAEJ2716_ENHANCED_SERIAL_FRAME_4_BIT_ID,
	SAEJ2716_ENHANCED_SERIAL_FRAME_8_BIT_ID,
	SAEJ2716_FAST_FRAME
};
```

congnguyenhuu commented on dts/bindings/saej2716/saej2716-controller.yaml:
```diff
@@ -0,0 +1,80 @@
+        base time unit for the system's timing operations, effectively setting the resolution
+        of the timer.
+
+    calib-method-low-latency:
```
I updated

congnguyenhuu commented on dts/bindings/saej2716/saej2716-controller.yaml:
```diff
@@ -0,0 +1,80 @@
+        Indicates whether the successive calibration pulse check method with low latency is enabled.
+        This method ensures faster calibration by reducing the delay between successive pulses.
+
+    calib-pulse-range-25:
```
I updated

congnguyenhuu commented on dts/bindings/saej2716/saej2716-controller.yaml:
```diff
@@ -0,0 +1,80 @@
+        Indicates whether a 25% variation in the calibration pulse is acceptable.
+        If not, a 20% variation is acceptable.
+
+    pause-pulse-enable:
```
I updated

congnguyenhuu commented:
I have updated to the v2 changes:
- combined property `crc-status-nibble-include` to property `fast-crc` due to this only apply for fast frame CRC
- Added require true for mandatory properties

congnguyenhuu commented:
I have updated to the v2 changes: added the check for the valid value of the binding properties `fast-crc` and `short-serial-crc`

congnguyenhuu commented:
I have updated to the v3 changes: added the check for the valid value of the binding properties `fast-crc` and `short-serial-crc`

Dat-NguyenDuy reviewed (dismissed):
Looks fine from my side

rruuaanng commented on MAINTAINERS.yml:
```diff
@@ -2039,6 +2039,23 @@ Release Notes:
   tests:
     - drivers.pwm
 
+"Drivers: SAEJ2716":
```
Why isn't it called SNET?

manuargue replied:
I got to admit that I'd rather prefer this driver being called SENT and use one of the proposals in https://github.com/zephyrproject-rtos/zephyr/pull/80117#discussion_r1973974912 using `sent_` as a prefix to address the ambiguous API names rather than using the standard name. It's not just better for readability but also for users to know SENT is supported in Zephyr. I've checked several SENT drivers provided by different vendors and all of them use SENT naming, e.g.
- https://software-dl.ti.com/C2000/docs/mcal/F29H85x/latest/src/Module_User_Guide_Cdd_Sent.html#cdd-sent-module
- https://www.st.com/resource/en/technical_note/tn1233-sent-demo-driver-for-the-spc57xxspc58xx-microcontroller-families-stmicroelectronics.pdf
- https://www.infineon.com/cms/en/tools/aurix-embedded-sw/AUTOSAR/Infineon/

manuargue commented:
hi @carlescufi , @henrikbrixandersen . Can we remove the architecture review label or is there anything else to discuss? Thanks

manuargue reviewed (changes requested):
use one of the proposals in https://github.com/zephyrproject-rtos/zephyr/pull/80117#discussion_r1973974912 using `sent_` as a prefix to address the ambiguous API names rather than using the standard name, more details in https://github.com/zephyrproject-rtos/zephyr/pull/80117#discussion_r2134957414. Propagate the change to other files as well.

congnguyenhuu commented:
> use one of the proposals in [#80117 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/80117#discussion_r1973974912) using `sent_` as a prefix to address the ambiguous API names rather than using the standard name, more details in [#80117 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/80117#discussion_r2134957414). Propagate the change to other files as well.

I updated to use `sent` instead of `saej2716`, there are changes in the root driver folder name, Kconfig and the API names
`saej2716_start_rx` to `sent_start_listening`
`saej2716_stop_rx` to `sent_stop_listening`
`saej2716_register_callback` to `sent_register_callback`

mmahadevan108 commented:
@henrikbrixandersen , can you take a quick peek at this PR.

congnguyenhuu commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,195 @@
+/**
+ * @brief SENT Interface
+ * @defgroup sent_interface SENT Interface
+ * @ingroup io_interfaces
```
I updated this

congnguyenhuu commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,202 @@
+		} serial;
+
+		struct {
+			uint8_t data_nibbles[SENT_MAX_DATA_NUBBLES];
```
I have changed the SENT fast frame structure to use an array of data nibbles instead of a single 32-bit data value.
And updated shim-driver and test, sample code to work with this change.

danieldegrasse commented on dts/bindings/sent/sent-controller.yaml:
```diff
@@ -0,0 +1,92 @@
+      required: true
+      description: Channel identifier.
+
+    num-data-nibbles:
```
Is this something we would want to configure at runtime, or will it only be configured at initialization?

congnguyenhuu replied:
As the specification recommends that SENT receiver allows for up to at least 8 data nibbles to ensure forward compatibility with future protocol revisions.

danieldegrasse commented on dts/bindings/sent/sent-controller.yaml:
```diff
@@ -0,0 +1,92 @@
+        Specifies the number of data nibbles. A nibble is a four-bit aggregation.
+        This property defines how many such nibbles are used in the data transmission.
+
+    clock-tick-length-us:
```
Same question- if both of these channel properties are set by the sender and static, than I think it makes sense to keep them in devicetree. Otherwise should there be an API to set these?

congnguyenhuu replied:
as my reply [above](https://github.com/zephyrproject-rtos/zephyr/pull/80117#discussion_r2165476748)

danieldegrasse commented on dts/bindings/sent/sent-controller.yaml:
```diff
@@ -0,0 +1,92 @@
+    fast-crc:
+      type: int
+      required: true
+      description: |
```
could we use an `enum` here, and use DT_ENUM_IDX in drivers?

congnguyenhuu replied:
The value of this property is a set of configuration flags that can be combined using OR. So, I use the int type for this.

danieldegrasse replied:
Ok, makes sense- fine with this then

danieldegrasse commented on dts/bindings/sent/sent-controller.yaml:
```diff
@@ -0,0 +1,92 @@
+    short-serial-crc:
+      type: int
+      required: true
+      description: |
```
Same question regarding enum. See here for an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mipi-dbi/mipi-dbi-device.yaml#L15

congnguyenhuu replied:
Although the value of this property is not a set of configuration flags like as `fast-crc`. But I use the int type for consistency across both `short-serial-crc` and `fast-crc` properties

danieldegrasse reviewed (commented):
Generally looks good to me- I'm not familiar with this protocol, so just a few clarifying questions on what we should define in devicetree versus runtime

kartben commented on tests/drivers/sent/sent_api/src/main.c:
```diff
@@ -0,0 +1,121 @@
+	int err;
+
+	err = sent_start_listening(sent_dev, SENT_CHANNEL);
+	zassert_equal(err, 0, "Failed to start rx (err %d)", err);
+
+	err = sent_start_listening(sent_dev, SENT_CHANNEL);
+	zassert_not_equal(err, 0, "Started rx while started");
```
I would suggest using z_assert_ok() / z_assert_not_ok() instead for all these kinds of asserts

congnguyenhuu commented on tests/drivers/sent/sent_api/src/main.c:
```diff
@@ -0,0 +1,121 @@
+	zassert_equal(err, -EALREADY, "Wrong error return code (err %d)", err);
+
+	err = sent_start_listening(sent_dev, SENT_CHANNEL);
+	zassert_equal(err, 0, "Failed to start rx (err %d)", err);
```
The SENT device is started during the test setup, so it is already running when the test case begins

congnguyenhuu commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,202 @@
+	};
+
+	uint32_t timestamp;
+	uint8_t crc;
```
I updated to add doxygen

congnguyenhuu commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,202 @@
+	sent_rx_frame_callback_t callback;
+	struct sent_frame *frame;
+	uint32_t max_num_frame;
+	void *user_data;
```
I updated

congnguyenhuu commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,202 @@
+
+struct sent_rx_callback_configs {
+	struct sent_rx_callback_config *serial;
+	struct sent_rx_callback_config *fast;
```
I updated

congnguyenhuu commented on include/zephyr/drivers/sent/sent.h:
```diff
@@ -0,0 +1,202 @@
+ * @retval 0 if successful.
+ * @retval -EINVAL if an invalid channel is given.
+ * @retval -EALREADY if the device is already started.
+ * @retval -EIO General input/output error, failed to start device.
```
I updated

congnguyenhuu commented on samples/drivers/sent/README.rst:
```diff
@@ -0,0 +1,31 @@
+Overview
+********
+
+The sample application shows how to use the Single Edge Nibble Transmission (SENT) driver:
```
I updated

congnguyenhuu commented:
twister-build is failing, it fixes by open [PR ](https://github.com/zephyrproject-rtos/zephyr/pull/92238)

kartben reviewed (approved):
Great work with the docs, much appreciated!

kartben reviewed (approved):
Great work with the docs, much appreciated!
